src/tree/node_text.rs

```diff
@@ -305,22 +305,11 @@ mod inner {
             // Enlarge buffer and shift bytes over to make room for the
             // insertion string.
             //
-            // There are two uses of unsafe here:
-            // - `set_len()`, which is used to avoid having to actually
-            //   initialize bytes that we know we're immediately going to
-            //   write to in the next step.
-            // - `ptr::copy()` to efficiently shift bytes over to make room
-            //   for the insertion data.  This can be replaced with a safe call
-            //   to `copy_within()` on the slice once that API is stabalized in
-            //   the standard library.
-            unsafe {
-                self.buffer.set_len(len + amt);
-                ptr::copy(
-                    self.buffer.as_ptr().add(byte_idx),
-                    self.buffer.as_mut_ptr().add(byte_idx + amt),
-                    len - byte_idx,
-                );
-            }
+            // We're using unsafe `set_len()` to avoid having to actually
+            // initialize bytes that we know we're immediately going to write
+            // to in the next step anyway.
+            unsafe { self.buffer.set_len(len + amt) };
+            self.buffer.copy_within(byte_idx..len, byte_idx + amt);
```

Comment:
Could most of these cases that use `set_len` just use `extend_from_slice` instead? We seem to `set_len + copy_within + copy_from_slice` which matches what `extend_from_slice` would do

Reply:
Since this is using smallvec it seems to do about the same: https://docs.rs/smallvec/latest/src/smallvec/lib.rs.html#1440-1467

Reply:
I don't believe we can use `extend_from_slice` here:

1. `extend_from_slice` takes a *different* slice as the source to copy from.  We can't pass the same slice as both the source and destination.
2. What we're doing isn't actually a strict extension, but often (usually, even) copying between overlapping ranges.  For example, we may be shifting a 200 byte chunk over by just a few bytes.

Reply:
This case seems like it could use it though  https://github.com/cessen/ropey/blob/98deaea340b809839b89a99c88189b266054cdfe/src/tree/node_text.rs#L358-L360

Reply:
Ah, okay.  So, we still can't use `extend_from_slice` because that only appends to the end, whereas we need to insert them in the middle.  But I just noticed that in smallvec code you linked to (which you highlighted, ha ha... I'm very unobservant sometimes) they also have `insert_from_slice`.  And we could use that, for sure.

Indeed, `insert_from_slice` is literally what I've implemented here, with the exception of the char-boundary check.  Feel free to make that change!